Sbt-hood plugin keys and main task

jdesiloniz opened the pull request:
Fixes: #12 

This PR introduces the main sbt task (`compareBenchmarks`) and configuration keys for sbt-hood.  This task compares the two benchmarks provided by the user and shows the comparison result in the console. All config keys are currently implemented (except for those related to GitHub/`compareBenchmarksCI` that are not part for the MVP).

It also fixes a couple of issues with error handling when it comes to load the files from disk.

fedefernandez commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,137 @@
+        currentBenchmarks
+          .get(benchmarkKey)
+          .fold({
+            L.error(
```
Potentially not executed, since it's inside a `delay` and is a side-effect

jdesiloniz replied:
Good catch! Please take a look at the fix I've submitted when you have a chance ;)

fedefernandez commented on modules/core/src/main/scala/com.fortysevendeg.hood/csv/CsvService.scala:
```diff
@@ -30,7 +30,12 @@ import scala.io.{BufferedSource, Source}
+      modeCol: String,
+      compareCol: String,
+      thresholdCol: String,
+      unitsCol: String)(csvFile: File): F[Either[HoodError, List[Benchmark]]]
```
I'd create a config `case class` to group all these values, makes sense?

jdesiloniz replied:
@fedefernandez Agree... fixed!

juanpedromoreno commented on modules/core/src/main/scala/com.fortysevendeg.hood/csv/CsvService.scala:
```diff
@@ -28,9 +28,17 @@ import io.chrisdavenport.log4cats.Logger
 
 import scala.io.{BufferedSource, Source}
 
+case class BenchmarkColumns(
```
```suggestion
final case class BenchmarkColumns(
```

juanpedromoreno commented on modules/core/src/main/scala/com.fortysevendeg.hood/csv/CsvService.scala:
```diff
@@ -28,9 +28,17 @@ import io.chrisdavenport.log4cats.Logger
 
-  def parseBenchmark(csvFile: File): F[Either[HoodError, List[Benchmark]]]
+  def parseBenchmark(columns: BenchmarkColumns)(
+      csvFile: File): F[Either[HoodError, List[Benchmark]]]
```
Could we use a single group of parameters?

juanpedromoreno commented on modules/core/src/main/scala/com.fortysevendeg.hood/benchmark/BenchmarkService.scala:
```diff
@@ -26,8 +28,17 @@ case object Error                      extends BenchmarkComparisonStatus
+    current: Option[Benchmark],
+    result: BenchmarkComparisonStatus,
+    threshold: Double) {
+  def icon(): String = {
```
```suggestion
  def icon: String = {
```

juanpedromoreno commented on modules/core/src/test/scala/com/fortysevendeg/hood/csv/CsvServiceTests.scala:
```diff
@@ -32,7 +32,12 @@ class CsvServiceTests extends FlatSpec with Matchers {
+      .parseBenchmark("Benchmark", "Mode", "Score", "Score Error (99.9%)", "Unit")(dataFile)
+      .unsafeRunSync()
+
+    println(s"Result: ${result}")
```
🔥

jdesiloniz replied:
Whoops!

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+      benchmarkThreshold.value
+    ).leftFlatMap(e =>
+        EitherT.left[List[BenchmarkComparisonResult]](logger.error(s"There was an error: $e")))
+      .value
```
.void

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+      .value
+      .unsafeRunSync()
+
+    ()
```
With `.void` (see comment above) you wouldn't need this

fedefernandez replied:
I guess we'd need a `.merge` + `.void`

jdesiloniz replied:
`.void` by itself seems to do the trick (without `.merge`), as in:

```
// ...
.leftFlatMap(e =>
        EitherT.left[List[BenchmarkComparisonResult]](logger.error(s"There was an error: $e")))
      .value
      .unsafeRunSync()
      .void
```

jdesiloniz replied:
Oh, now I see what you mean: without `merge` the compiler complains about discarding a value (because `.void` converts the result to `Either(())`). `.void.merge` fixes this. Thanks!!

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+      implicit L: Logger[F],
+      S: Sync[F],
+      C: Console[F],
+      M: Monad[F]): EitherT[F, HoodError, List[BenchmarkComparisonResult]] = {
```
With `Sync` you wouldn't need `Monad`

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+      thresholdMap: Map[String, Double])(
+      implicit L: Logger[F],
+      S: Sync[F],
+      M: Monad[F]): F[List[BenchmarkComparisonResult]] =
```
Same here

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+      S: Sync[F],
+      M: Monad[F]): F[List[BenchmarkComparisonResult]] =
+    previousBenchmarks
+      .map {
```
I think you could swap this `map` by `traverse` and remove the `.sequence` at the end

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,146 @@
+         |$currentFile|${result.current.map(_.primaryMetric.score.toString).getOrElse("N/A")}
+    """.stripMargin
+
+    benchmarks.map(b => outputComparisonResult(b)).mkString("")
```
```suggestion
    benchmarks.map(outputComparisonResult).mkString("")
```

juanpedromoreno commented on modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala:
```diff
@@ -0,0 +1,141 @@
+
+  def compareBenchmarksTask: Def.Initialize[Task[Unit]] = Def.task {
+
+    implicit val logger = Slf4jLogger.getLogger[IO]
```
We could initialize this logger as part of the plugin initialization. We may just move it at object level.

jdesiloniz commented:
@fedefernandez @juanpedromoreno 👋 Thanks for the suggestions! I think I've addressed all your comments. Could you please review again when you have a chance? Thanks!!
